osquery/tables/yara/windows/yara_events.cpp

```diff
@@ -0,0 +1,214 @@
+using FileEventContextRef = NTFSEventContextRef;
+using FileSubscriptionContextRef = NTFSEventSubscriptionContextRef;
+
+/**
+ * @brief Track YARA matches to files.
+ */
+class YARAEventSubscriber : public FileEventSubscriber {
+ public:
+  Status init() override {
+    return Status::success();
+  }
+
+  void configure() override;
+
+ private:
+  /**
+   * @brief This exports a single Callback for FSEventsEventPublisher events.
+   *
+   * @param event The Callback type receives an EventContextRef substruct
+   * for the FSEventsEventPublisher declared in this EventSubscriber subclass.
+   *
+   * @return Status
+   */
+  Status Callback(const FileEventContextRef& event,
+                  const FileSubscriptionContextRef& sc);
+};
+
+/**
+ * @brief Each EventSubscriber must register itself so the init method is
+ * called.
+ *
+ * This registers YARAEventSubscriber into the osquery EventSubscriber
+ * pseudo-plugin registry.
+ */
+REGISTER(YARAEventSubscriber, "event_subscriber", "yara_events");
+
+void YARAEventSubscriber::configure() {
+  removeSubscriptions();
+
+  // There is a special yara parser that tracks the related top-level keys.
+  auto plugin = Config::getParser("yara");
+  if (plugin == nullptr || plugin.get() == nullptr) {
+    return;
+  }
+
+  // Bail if there is no configured set of opt-in paths for yara.
+  const auto& yara_config = plugin->getData().doc();
+  if (!yara_config.HasMember("file_paths") ||
+      !yara_config["file_paths"].IsObject()) {
+    return;
+  }
+
+  const auto& json = Config::getParser("file_paths")->getData();
+  const auto& json_document = json.doc();
+
+  // Collect the set of paths, we are mostly concerned with the categories.
+  // But the subscriber must duplicate the set of subscriptions such that the
+  // publisher's 'fire'-matching logic routes related events to our callback.
+  std::map<std::string, std::vector<std::string>> file_map;
+  Config::get().files(
+      [this, &file_map, &json_document](const std::string& category,
+                                        const std::vector<std::string>& files) {
+        file_map[category] = files;
+      });
+
+  // For each category within yara's file_paths, add a subscription to the
+  // corresponding set of paths.
+  const auto& yara_paths = yara_config["file_paths"];
+  for (const auto& yara_path_element : yara_paths.GetObject()) {
+    std::string category = yara_path_element.name.GetString();
+    // Subscribe to each file for the given key (category).
+    if (file_map.count(category) == 0) {
+      VLOG(1) << "Key in YARA file_paths not found in file_paths: " << category;
+      continue;
+    }
+
+    StringList include_path_list = {};
+    for (const auto& file : file_map.at(category)) {
+      // NOTE(woodruffw): This will remove nonexistent paths, even if
+      // they aren't patterns. For example, C:\foo\bar won't
+      // be monitored if it doesn't already exist at table/event
+      // creation time. Is that what we want?

```

Comment:
mmmm. That is a good question. My hunch is that we want to let people monitor things that don't exist yet, but I don't know if that's really possible in the bounds of FIM

Reply:
Agree. This is copied from the `ntfs_journal_events` table and so shares the same behavior.